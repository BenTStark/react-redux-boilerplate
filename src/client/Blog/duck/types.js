const GET_BLOG_ARTICLES = "GET_BLOG_ARTICLES";
const CREATE_BLOG_ARTICLE = "CREATE_BLOG_ARTICLE";
const DELETE_BLOG_ARTICLE = "DELETE_BLOG_ARTICLE";
const SET_CURRENT_ARTICLE = "SET_CURRENT_ARTICLE";
const CLEAR_CURRENT_ARTICLE = "CLEAR_CURRENT_ARTICLE";
const EDIT_BLOG_ARTICLE = "EDIT_BLOG_ARTICLE";
const FINISH_EDIT_BLOG_ARTICLE = "FINISH_EDIT_BLOG_ARTICLE";
const UPDATE_BLOG_ARTICLE = "UPDATE_BLOG_ARTICLE";
const UPDATE_EDITOR_STATE = "UPDATE_EDITOR_STATE";

export default {
  GET_BLOG_ARTICLES,
  CREATE_BLOG_ARTICLE,
  DELETE_BLOG_ARTICLE,
  SET_CURRENT_ARTICLE,
  CLEAR_CURRENT_ARTICLE,
  EDIT_BLOG_ARTICLE,
  FINISH_EDIT_BLOG_ARTICLE,
  UPDATE_BLOG_ARTICLE,
  UPDATE_EDITOR_STATE
};
