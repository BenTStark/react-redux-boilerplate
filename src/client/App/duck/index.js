import authReducer from './reducers';
export default authReducer;
export { AppOperations } from './operations';
export { default as appTypes } from './types';
