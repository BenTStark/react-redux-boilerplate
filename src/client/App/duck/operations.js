// optional for API requests
import ActionCreators from "./actions";
import AuthService from "./auth.service";
import types from "./types";

const loginRequest = () => ActionCreators.loginRequest();
const loginSuccess = authInformation => ActionCreators.loginSuccess(authInformation);
const loginError = () => ActionCreators.loginError();
const logout = () => ActionCreators.logout();

const checkLogin = () => {
  return new Promise((resolve, reject) => {
    const payload = {},
      authInformation = {};
    authInformation.result = "";
    if (AuthService.loggedIn()) {
      payload.profile = AuthService.getProfile();
      payload.accessToken = AuthService.getAccessToken();
      authInformation.result = types.LOGIN_SUCCESS;
    }
    authInformation.payload = payload;
    resolve(authInformation);
  });
};

const authentication = () => {
  // Add callback for lock's `authenticated` event
  return new Promise((resolve, reject) => {
    const payload = {},
      authInformation = {};
    authInformation.result = "";
    authInformation.payload = payload;
    AuthService.lock.on("authenticated", authResult => {
      AuthService.lock.getUserInfo(authResult.accessToken, (error, profile) => {
        if (error) {
          authInformation.result = types.LOGIN_ERROR;
          reject(authInformation);
        } else {
          AuthService.setToken(authResult.idToken, authResult.accessToken); // static method
          AuthService.setProfile(profile); // static method
          payload.profile = profile;
          payload.accessToken = authResult.accessToken;
          authInformation.result = types.LOGIN_SUCCESS;
          authInformation.payload = payload;
          AuthService.lock.hide();
          resolve(authInformation);
        }
      });
    });
    // Add callback for lock's `authorization_error` event
    AuthService.lock.on("authorization_error", error => {
      authInformation.result = types.LOGIN_ERROR;
      reject(authInformation);
    });
  });
};

const handleLogin = () => {
  AuthService.login();
};

const handleLogout = () => {
  AuthService.logout();
};

export const AppOperations = {
  handleLogin,
  loginRequest,
  loginSuccess,
  loginError,
  handleLogout,
  logout,
  checkLogin,
  authentication
};

//export default AppOperations;
