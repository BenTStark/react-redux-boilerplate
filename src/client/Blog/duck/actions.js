import types from "./types";

// Actions should not do any change to the value they pass to the store.
// Any value treatment should happen before in operations.js!
// This is an example how an action should look like:
const getBlogArticles = articles => {
  return {
    type: types.GET_BLOG_ARTICLES,
    articles: articles
  };
};

const createBlogArticle = (articleId, article) => {
  return {
    type: types.CREATE_BLOG_ARTICLE,
    articleId: articleId,
    article: article
  };
};

const deleteBlogArticle = articleId => {
  return {
    type: types.DELETE_BLOG_ARTICLE,
    articleId: articleId
  };
};

const setCurrentArticle = article => {
  return {
    type: types.SET_CURRENT_ARTICLE,
    article: article
  };
};

const clearCurrentArticle = () => {
  return {
    type: types.CLEAR_CURRENT_ARTICLE
  };
};

const editBlogArticle = articleId => {
  return {
    type: types.EDIT_BLOG_ARTICLE,
    articleId: articleId
  };
};

const finishEditBlogArticle = articleId => {
  return {
    type: types.FINISH_EDIT_BLOG_ARTICLE,
    articleId: articleId
  };
};

const updateBlogArticle = (articleId, article) => {
  return {
    type: types.UPDATE_BLOG_ARTICLE,
    articleId: articleId,
    article: article
  };
};

const updateEditorState = editorState => {
  return {
    type: types.UPDATE_EDITOR_STATE,
    payload: editorState
  };
};

export default {
  getBlogArticles,
  createBlogArticle,
  deleteBlogArticle,
  setCurrentArticle,
  clearCurrentArticle,
  editBlogArticle,
  finishEditBlogArticle,
  updateBlogArticle,
  updateEditorState
};
