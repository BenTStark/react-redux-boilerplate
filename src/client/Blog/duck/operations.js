// optional for API requests
import ActionCreators from "./actions";
import _ from "lodash";
import moment from "moment";

const getDataUrl = "https://mirrorpi.ddns.net/cms/blogs";
const updateDataUrl = articleId =>
  `https://mirrorpi.ddns.net/cms/blogs/${articleId}`;

const getBlogArticles = articles => ActionCreators.getBlogArticles(articles);
const createBlogArticle = (articleId, article) => {
  return ActionCreators.createBlogArticle(articleId, article);
};
const deleteBlogArticle = articleId =>
  ActionCreators.deleteBlogArticle(articleId);
const setCurrentArticle = article => ActionCreators.setCurrentArticle(article);
const clearCurrentArticle = () => ActionCreators.clearCurrentArticle();
const editBlogArticle = articleId => ActionCreators.editBlogArticle(articleId);
const finishEditBlogArticle = articleId =>
  ActionCreators.finishEditBlogArticle(articleId);
const updateBlogArticle = (articleId, article) =>
  ActionCreators.updateBlogArticle(articleId, article);

const buildEditorContent = article => {
  return [
    ["title:", article.title + ","].join(" "),
    ["author:", article.author].join(" "),
    "~~~",
    article.text
  ].join("\r\n");
};
const buildArticleObject = plainText => {
  // Search for these regex in plainText to find the respective values
  const keys = [
    { key: "author", regex: /author/ },
    { key: "title", regex: /title/ }
  ];
  // target object should have keys for the article object in BlogComponent
  const result = {};
  const arr = plainText.split("~~~");
  arr[0].split(",").forEach(element => {
    keys.forEach(key => {
      if (key.regex.test(element.split(":")[0])) {
        result[key.key] = element.split(":")[1].trim();
      }
    });
  });
  // Remove \r \n
  result.text = arr[1].substring(1);
  result.isNew = false;
  return result;
};

const maxId = array => {
  return array.length ? Math.max(...array) : 1;
};

const sortBlogArticle = (a, b, articles) => {
  if (
    _.gt(
      moment.utc(articles[a].createdAt).unix(),
      moment.utc(articles[b].createdAt).unix()
    )
  ) {
    return -1;
  }
  if (
    _.lt(
      moment.utc(articles[a].createdAt).unix(),
      moment.utc(articles[b].createdAt).unix()
    )
  ) {
    return 1;
  }
  return 0;
};

export const BlogOperations = {
  getBlogArticles,
  getDataUrl,
  updateDataUrl,
  createBlogArticle,
  deleteBlogArticle,
  setCurrentArticle,
  clearCurrentArticle,
  editBlogArticle,
  finishEditBlogArticle,
  updateBlogArticle,
  buildEditorContent,
  buildArticleObject,
  maxId,
  sortBlogArticle
};
