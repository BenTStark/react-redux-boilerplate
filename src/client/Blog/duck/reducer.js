import { EditorState } from "draft-js";
import types from "./types";
import _ from "lodash";
import moment from "moment";

const initialState = {
  byId: [],
  byHash: {},
  editorState: EditorState.createEmpty(),
  currentArticle: {
    title: null,
    createdAt: null,
    modifiedAt: null,
    author: null,
    text: null,
    editMode: false,
    isNew: false
  }
};

function getBlogArticles(state, articles) {
  const ids = articles.map(article => {
    return article.id;
  });
  const newArticles = {};
  articles.forEach(article => {
    newArticles[article.id] = {
      title: article.title,
      createdAt: article.createdAt,
      modifiedAt: article.modifiedAt,
      author: article.author,
      text: article.text,
      editMode: false,
      isNew: false
    };
  });
  return {
    ...state,
    byId: state.byId.concat(ids.filter(id => !state.byId.includes(id))),
    byHash: {
      ...state.byHash,
      ...newArticles
    }
  };
}

function createBlogArticle(state, articleId, article) {
  return {
    ...state,
    byId: [...state.byId, articleId],
    byHash: {
      ...state.byHash,
      [articleId]: { ...article, isNew: true }
    }
  };
}

function deleteBlogArticle(state, id) {
  return {
    ...state,
    byId: state.byId.filter(existingId => existingId !== id),
    byHash: _.omit(state.byHash, [id])
  };
}

function setCurrentArticle(state, article) {
  return {
    ...state,
    currentArticle: article
  };
}

function clearCurrentArticle(state) {
  return {
    ...state,
    currentArticle: state.currentArticle
  };
}

function editBlogArticle(state, id) {
  return {
    ...state,
    byHash: {
      ...state.byHash,
      [id]: {
        ...state.byHash[id],
        editMode: true
      }
    }
  };
}

function finishEditBlogArticle(state, id) {
  return {
    ...state,
    byHash: {
      ...state.byHash,
      [id]: {
        ...state.byHash[id],
        editMode: false
      }
    }
  };
}

function updateBlogArticle(state, articleId, article) {
  return {
    ...state,
    byHash: {
      ...state.byHash,
      [articleId]: article
    }
  };
}

function updateEditorState(state, payload) {
  return {
    ...state,
    editorState: payload
  };
}

const blogReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.GET_BLOG_ARTICLES:
      return getBlogArticles(state, action.articles);
    case types.CREATE_BLOG_ARTICLE:
      return createBlogArticle(state, action.articleId, action.article);
    case types.DELETE_BLOG_ARTICLE:
      return deleteBlogArticle(state, action.articleId);
    case types.SET_CURRENT_ARTICLE:
      return setCurrentArticle(state, action.article);
    case types.CLEAR_CURRENT_ARTICLE:
      return clearCurrentArticle(state);
    case types.EDIT_BLOG_ARTICLE:
      return editBlogArticle(state, action.articleId);
    case types.FINISH_EDIT_BLOG_ARTICLE:
      return finishEditBlogArticle(state, action.articleId);
    case types.UPDATE_BLOG_ARTICLE:
      return updateBlogArticle(state, action.articleId, action.article);
    case types.UPDATE_EDITOR_STATE:
      return updateEditorState(state, action.payload);
    default:
      return state;
  }
};

export default blogReducer;
