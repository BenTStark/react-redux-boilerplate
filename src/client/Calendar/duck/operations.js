// optional for API requests
import ActionCreators from "./actions";


// This is an example how a simple operation could look like:
const defaultOperation = value => ActionCreators.defaultAction(value);





export const CalendarOperations = {
  defaultOperation
};
