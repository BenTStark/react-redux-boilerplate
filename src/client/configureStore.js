import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { routerMiddleware } from "connected-react-router";
import reducer from "./reducer";

export default function configureStore(history, initialState = {}) {
  const reduxRouterMiddleware = routerMiddleware(history);
  const middleware = [thunk, reduxRouterMiddleware];

  const enhancer = compose(
    // Middleware you want to use in development:
    applyMiddleware(thunk, reduxRouterMiddleware),
    // Required! Enable Redux DevTools with the monitors you chose
    window.devToolsExtension ? window.devToolsExtension() : f => f
  );

  // Sync dispatched route actions to the history

  const store = createStore(reducer(history), initialState, enhancer);

  // if (module.hot) {
  //   module.hot.accept('../reducers', () => {
  //     store.replaceReducer(require('./../reducers').default);
  //   });
  // }
  return store;
}
