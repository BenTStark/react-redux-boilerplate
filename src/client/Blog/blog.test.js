import BlogComponent from "./blog.component";
// To Do: bei App Template wird das nicht genutzt (weils aber auch nur ein Component ist)
import BlogContainer from "./blog.container";
import { BlogOperations } from "./duck/operations";

import React from "react";
import { shallow, mount } from "enzyme";
import configureStore from "redux-mock-store";
import toJson from "enzyme-to-json";
import MockAdapter from "axios-mock-adapter";

const blogProps = {
  byId: [1],
  byHash: {
    "1": {
      title: "title",
      createdAt: "1970.01.01",
      modifiedAt: "1970.01.01",
      author: "author",
      text: "text",
      editMode: false
    }
  }
};
const mockGetBlogArticlesfn = jest.fn(() => {
  return [
    {
      id: 2,
      title: "title",
      createdAt: "1970-01-01",
      modifiedAt: "1970-01-01",
      author: "author",
      text: "text"
    }
  ];
});

//Snapshot for Blog React Component
describe(">>> BlogComponent - Snapshot", () => {
  it("+++capturing Snapshot of BlogComponent", () => {
    // To Do: bei App Template wird das nicht genutzt wird hier shallow genutzt!
    const renderedValue = mount(
      <BlogComponent blog={blogProps} getBlogArticles={mockGetBlogArticlesfn} />
    );
    expect(toJson(renderedValue)).toMatchSnapshot();
  });
});

// Component Test
describe(">>> BlogComponent --- Shallow Render React Components", () => {
  const wrapper = shallow(
    <BlogComponent blog={blogProps} getBlogArticles={mockGetBlogArticlesfn} />
  );

  it("+++ render the COMPONENT", () => {
    expect(wrapper.length).toEqual(1);
  });
  it("+++ getBlogArticles Function working", () => {
    expect(mockGetBlogArticlesfn.mock.calls.length).toBeGreaterThan(1);
  });
});

// Container Test
describe(">>> BlogContainer --- REACT-REDUX (Shallow + passing the {store} directly)", () => {
  const initialState = {
    blogReducer: blogProps
  };
  const mockStore = configureStore();
  let store, container;

  beforeEach(() => {
    store = mockStore(initialState);
    container = shallow(
      <BlogContainer store={store} getBlogArticles={mockGetBlogArticlesfn} />
    );
  });

  it("+++ check Prop matches with initialState", () => {
    expect(container.prop("blog")).toEqual(initialState.blogReducer);
  });
});
