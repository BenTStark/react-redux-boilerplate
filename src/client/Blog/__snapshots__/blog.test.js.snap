// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> BlogComponent - Snapshot +++capturing Snapshot of BlogComponent 1`] = `
<BlogComponent
  blog={
    Object {
      "byHash": Object {
        "1": Object {
          "author": "author",
          "createdAt": "1970.01.01",
          "editMode": false,
          "modifiedAt": "1970.01.01",
          "text": "text",
          "title": "title",
        },
      },
      "byId": Array [
        1,
      ],
    }
  }
  getBlogArticles={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Array [
            Object {
              "author": "author",
              "createdAt": "1970-01-01",
              "id": 2,
              "modifiedAt": "1970-01-01",
              "text": "text",
              "title": "title",
            },
          ],
        },
        Object {
          "isThrow": false,
          "value": Array [
            Object {
              "author": "author",
              "createdAt": "1970-01-01",
              "id": 2,
              "modifiedAt": "1970-01-01",
              "text": "text",
              "title": "title",
            },
          ],
        },
      ],
    }
  }
>
  <div
    className="blog"
  >
    <div>
      <h1>
        Blog
      </h1>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add
      </button>
      <button
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Test
      </button>
      <div
        key="1"
      >
        <div>
          <h2>
            title
          </h2>
          <span>
            by 
            author
          </span>
          <button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Edit
          </button>
          <ReactMarkdown
            astPlugins={Array []}
            escapeHtml={true}
            plugins={Array []}
            rawSourcePos={false}
            renderers={Object {}}
            skipHtml={false}
            source="text"
            sourcePos={false}
            transformLinkUri={[Function]}
          >
            <Root
              key="root-1-1"
            >
              <p
                key="paragraph-1-1"
              >
                <TextRenderer
                  key="text-1-1"
                  nodeKey="text-1-1"
                  value="text"
                >
                  text
                </TextRenderer>
              </p>
            </Root>
          </ReactMarkdown>
        </div>
      </div>
    </div>
  </div>
</BlogComponent>
`;
